test/fixtures/es-module-loaders/hook-resolve-type.mjs

```diff
@@ -0,0 +1,19 @@
+let importedESM = 0;
+let importedCJS = 0;
+global.getModuleTypeStats = () => { return {importedESM, importedCJS} };
+
+export function load(url, context, next) {
+  return next(url, context, next);
+}
+
+export function resolve(specifier, context, next) {
+  const { url, format } = next(specifier, context);
+
+  if ((format != null && format == 'module') || specifier.endsWith('.mjs')) {
+    importedESM++;
+  } else if (format == null || format == 'commonjs') {
+    importedCJS++;
+  } 
+
+  return {url, ...(format != null) && {format} };
```

Comment:
Can't we simplify that?
```suggestion
  return { url, format };
```

Reply:
even if it only is a subtle difference for the case where format is not resolved by the defaultResolver I would still let it as it is. 
On the return object { url, format } for example hasOwnProperty('format') will answer true even if format is not returned by the defaultresolver. with the conditional as it currently is, hasOwnProperty('format') will return false hence keeping the behaviour of the default.

Reply:
I think it's a valid argument for code in `lib/`, but in this test it will never matter because we don't call `Object.hasOwn` on the result. I would still remove it here.

Reply:
writing (_slightly_) less correct code just because it looks easier in syntax is imho not something we as a community should strive to achieve.
Besides, writing a custom loader is not part of the "hello world" Node.js example so it can be expected that people looking at these tests to find an example will understand that syntax immediatelly. I added a short _green text_ to explain why we chose the conditional spread syntax there

Reply:
How is it less correct to pass `format: undefined`?

Reply:
One of the goals of the additional loader is to wrap the default loader and produce no observable side-effects. The fact that `hasOwnProperty('format')` returns different results if applied on the returned value of custom loader `resolve` and on original default `resolve` *is* an observable side-effect of introducing the additional loader. The fact that this is currently not used does not exclude the possibility that `lib\internal\modules\esm\loader.js` (or its replacement) will use it in the future.
I know it is theoretical but it does not make it less true. In fact I think we are overdiscussing it as it will not be of too much importance since we're talking of an idle sample code. Nonetheless, it is as I see it an argument for production-level code.
So if the condition to get the PR through is to return `{ url, format }` in the sample loader just let me know and I will do that commit.

Reply:
To be clear, this is not a blocking concern from me (this PR is only blocked by https://github.com/nodejs/node/pull/41132 as far as I'm concerned); I'm discussing this because I would be interested in writing loaders myself, to understand how to write them correctly.

> The fact that `hasOwnProperty('format')` returns different results if applied on the returned value of custom loader `resolve` and on original default `resolve` _is_ an observable side-effect of introducing the additional loader.


It would probably be simpler to return the original object then.

Alternative proposal: make the default loader always return an object with a `format` property, so `Object.hasOwn('format')` is always `true`. It's not uncommon is ECMAScript to return an object with a property whose value is `undefined`, e.g.:
```js
const obj = {};
Object.defineProperty(obj, 'property', { get() { return 4 } });
console.log(Object.hasOwn(Reflect.getOwnPropertyDescriptor(obj, 'property'), 'set')); // true
```

What do you think?

Reply:
As a matter of fact my initial take on this would have been to always try to resolve the type based on the `package.json`. That means we have the defaultResolver *always* return the `format` as good as it can tell. But that breaks existing tests, it possibly creates other side-effects. Actually that was also a review comment from @bmeck on this PR: https://github.com/nodejs/node/pull/40980#discussion_r758607561 
Because this would be my first contribution to the project I decided to take it one step at a time. Only where currently existing code has this information, pass it out of the function. This keeps existing tests green and allows external loaders to benefit of the information computed by `defaultResolve` in some cases. This is a minimally invasive change.
If I were to choose, I would try to close this up one way or the other. If it gets merged, next steps can be discussed to go in the direction of always returning `format` from `defaultResolve`. But that comes with an interface change, documentation update, and so on. This also may have some impact on ongoing loaders group work as far as I can tell from where I stay. So I'm not sure that my guess is enough of an "educated guess" and for sure I'm not the one to make this decision. If the whole loaders mechanism gets reimplemented from scratch based on the chaining redesign and considered obsolete, it might be that it's just water down the drain.
After https://github.com/nodejs/node/pull/41132 gets merged, I can rebase and squash, remove the `'url'` import and we take it from there. Would that make sense?

Reply:
OK that makes sense. Would you agree https://github.com/nodejs/node/pull/40980#discussion_r766752811 simplify the code?

> After #41132 gets merged, I can rebase and squash, remove the `'url'` import and we take it from there. Would that make sense?

Yes that would be perfect (FYI you don't have to squash yourself, but if you prefer to do it that's fine by me), then we can run the CI and land this finally :)

Reply:
> Would you agree [#40980 (comment)](https://github.com/nodejs/node/pull/40980#discussion_r766752811) simplify the code?

Yes, I will take this over with the next commit. Will do it with the rebase after your `fs` bugfix gets merged.